### PR TITLE
Bluetooth: Mesh: Enhancement prov segment received

### PR DESCRIPTION
**Pull request by LingaoM**
Optimize provisioning segment package received, and
not ignore non-begin segment.

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

### CODE REVIEW

File: subsys/bluetooth/mesh/pb_adv.c [hunk 1]
```diff
@@ -412,12 +417,6 @@ static void gen_prov_start(struct prov_rx *rx, struct net_buf_simple *buf)
 		return;
 	}
 
-	if (rx->xact_id != expected_id) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/pb_adv.c:415]: This check has disappeared from this function completely. Should be reintroduced, along with a duplication check, for instance in combination with the check on line 411, like:

```C
if (rx->xact_id == link.rx.id) {
	if (!link.rx.seg) {
		if (!ack_pending()) {
			BT_DBG("Resending ack");
			gen_prov_ack_send(rx->xact_id);
		}

		return;
	}

	if (!(link.rx.seg & BIT(0))) {
		BT_DBG("Ignoring duplicate segment");
		return;
	}
} else if (rx->xact_id != next_transaction_id(link.rx.id)) {
	BT_WARN("Unexpected xact 0x%x, expected 0x%x", rx->xact_id,
		next_transaction_id(link.rx.id));
	return;
}
```

File: subsys/bluetooth/mesh/pb_adv.c [hunk 2]
```diff
@@ -445,16 +444,46 @@ static void gen_prov_start(struct prov_rx *rx, struct net_buf_simple *buf)
...
+			prov_failed(PROV_ERR_NVAL_FMT);
+			return;
+		}
+	}
+
+	if (link.rx.seg) {
+		seg = link.rx.seg;
+	}
+
+	link.rx.seg = (seg & (1 << (START_LAST_SEG(rx->gpc) + 1))) - 1;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/pb_adv.c:465]: This doesn't look quite right. Shouldn't it be 
```C
link.rx.seg = seg & ((1 << (START_LAST_SEG(rx->gpc) + 1)) - 1);
```
as `seg` is already a bitmask?

File: subsys/bluetooth/mesh/pb_adv.c [hunk 3]
```diff
@@ -445,16 +444,46 @@ static void gen_prov_start(struct prov_rx *rx, struct net_buf_simple *buf)
...
+	if (link.rx.seg) {
+		return;
 	}
+
+	if (link.rx.last_seg) {
+		uint8_t expect_len;
+
+		expect_len = (link.rx.buf->len - 20U -
+			      ((link.rx.last_seg - 1) * 23U));
+		if (expect_len != link.rx.buf->len) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/pb_adv.c:478]: This check will always fail. 

If we receive the last segment before the start segment, there's no way to perform this check with the current design. The `link.rx.buf->len` would need a different approach to work for all scenarios.

File: subsys/bluetooth/mesh/pb_adv.c [hunk 4]
```diff
@@ -445,16 +444,46 @@ static void gen_prov_start(struct prov_rx *rx, struct net_buf_simple *buf)
...
 	memcpy(link.rx.buf->data, buf->data, buf->len);
 	XACT_SEG_RECV(0);
 
-	if (!link.rx.seg) {
-		prov_msg_recv();
+	if (link.rx.seg) {
+		return;
 	}
+
+	if (link.rx.last_seg) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/pb_adv.c:473]: Should perform a check like this even for single segment messages

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Thanks for addressing this.

I have some comments that need to be resolved before this works as intended.

File: subsys/bluetooth/mesh/pb_adv.c [hunk 5]
```diff
@@ -447,8 +459,20 @@ static void gen_prov_start(struct prov_rx *rx, struct net_buf_simple *buf)
 
 	prov_clear_tx();
 
-	link.rx.seg = (1 << (START_LAST_SEG(rx->gpc) + 1)) - 1;
 	link.rx.last_seg = START_LAST_SEG(rx->gpc);
+
+	/* Confirm that none of the received continuation segments are out of bounds: */
+	if ((link.rx.seg & BIT(0)) && (find_msb_set(~link.rx.seg) >= link.rx.last_seg) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/pb_adv.c:465]: I'm afraid I made some compiler errors for you in my snippet. There's a missing closing parenthesis here

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
PR looks good, but please resolve the compliance check errors: https://github.com/zephyrproject-rtos/zephyr/pull/27226/checks?check_run_id=951838318

**Review by trond-snekvik — APPROVED** [review]

File: subsys/bluetooth/mesh/pb_adv.c [hunk 6]
```diff
@@ -334,27 +335,28 @@ static void gen_prov_cont(struct prov_rx *rx, struct net_buf_simple *buf)
 		return;
 	}
 
-	if (rx->xact_id != link.rx.id) {
+	if (!link.rx.seg &&
+		rx->xact_id == next_transaction_id(link.rx.id)) {
```
Comment (joerchan) [inline subsys/bluetooth/mesh/pb_adv.c:339]: Misleading indentation. This line should be aligned on the with the content of the if-statement.

**Comment by LingaoM** [comment]
@joerchan @jhedberg CC

**Review by jhedberg — APPROVED** [review]

**Pull request merged**